src/Test/StateMachine/Internal/Parallel.hs

```diff
@@ -239,3 +359,93 @@ toBoxDrawings prog = toBoxDrawings' allVars
 
       evT :: [(EventType, Pid)]
       evT = toEventType (filter (\e -> getProcessIdEvent e `elem` map Pid [1,2]) h)
+
+toBoxDrawings' :: HFoldable act => ParallelProgram' act -> History act err -> Doc
+toBoxDrawings' (ParallelProgram' prefix suffixes) = toBoxDrawings'' allVars
+  where
+  allVars = usedVars prefix `S.union` foldMap usedVars suffixes
+
+  toBoxDrawings'' :: Set Var -> History act err -> Doc
+  toBoxDrawings'' knownVars (History h) = exec evT (fmap out <$> Fork l p r)
+    where
+      (p, h') = partition (\e -> getProcessIdEvent e == Pid 0) h
+      (l, r)  = partition (\e -> getProcessIdEvent e == Pid 1) h'
+
+      out :: HistoryEvent act err -> String
+      out (InvocationEvent _ str var _)
+        | var `S.member` knownVars = show var ++ " ← " ++ str
+        | otherwise = str
+      out (ResponseEvent _ str _) = str
+
+      toEventType :: [HistoryEvent act err] -> [(EventType, Pid)]
+      toEventType = map go
+        where
+        go e = case e of
+          InvocationEvent _ _ _ pid -> (Open,  pid)
+          ResponseEvent   _ _   pid -> (Close, pid)
+
+      evT :: [(EventType, Pid)]
+      evT = toEventType (filter (\e -> getProcessIdEvent e `elem` map Pid [1,2]) h)
+
+------------------------------------------------------------------------
+
+splitProgram
+  :: Precondition model act
+  -> Transition'  model act err
+  -> model Symbolic
+  -> Program act
+  -> [Program act]
+splitProgram precondition transition model0 = go model0 [] . unProgram
+  where
+  go _     acc []    = reverse acc
+  go model acc iacts = go (advanceModel transition model (Program safe)) (Program safe : acc) rest
+    where
+    (safe, rest) = spanSafe model [] iacts
+
+  spanSafe _     safe []                            = (reverse safe, [])
+  spanSafe model safe (iact@(Internal _ _) : iacts)
+    | length safe <= 5 && parallelSafe precondition transition model (Program (iact : safe))
+        = spanSafe model (iact : safe) iacts
+    | otherwise
+        = (reverse safe, iact : iacts)
+
+parallelSafe
+  :: Precondition model act
+  -> Transition' model act err
+  -> model Symbolic
+  -> Program act
+  -> Bool
+parallelSafe precondition transition model0
+  = and
+  . map (preconditionsHold model0)
+  . permutations
```

Comment:
This doesn't seem right. Doesn't each thread run in program order?

Reply:
My understanding is that `parallelSafe` is testing just one thread at a time.

Reply:
Something that's perhaps a bit confusing is that the list of programs (suffixes) in: `ParallelProgram' act = ParallelProgram' (Program act) [Program act]` isn't of length 2 (for two threads)... The idea is that every program (or list of internal actions) in that list can be run in parallel. Perhaps the type should be `ParallelProgram' act = ParallelProgram' (Program act) [[Internal act]]` and a comment added that says that the inner list is parallel safe to make things more clear...

Reply:
By the way, I thought about how to generate only two suffixes (one per worker), but I couldn't come up with a good way (because I agree that it would be more intuitive that way). Ideas?

Reply:
Each suffix is a program for one worker which runs it in sequence, and workers can be run in parallel. But here it seems we are permuting the actions of one worker.

Reply:
My understanding is that each program in the suffix list gets split up between the workers upon execution (again I agree that this isn't intuitive given the type), and therefore this check makes sense!? :-)

Reply:
Ah! I got unstuck from thinking the `[ParallelProgram']` is the transpose of what it really is. Now it makes sense...

Reply:
Good :-). 

@Danten suggested moving the splitting from the point of execution to the type itself, something like: `ParallelProgram act = ParallelProgram (Program act) [(Program act, Program act)]` to make things more clear...

Reply:
@Lysxia: Changed the type around, do you find it easier to read now?

Reply:
Yes indeed :+1: